cmd/akamai-purger/main.go

```diff
@@ -193,36 +194,54 @@ func (ap *akamaiPurger) len() int {
 
 func (ap *akamaiPurger) purge() error {
 	ap.Lock()
-	queueEntries := ap.toPurge[:]
-	ap.toPurge = [][]string{}
-	ap.Unlock()
-	if len(queueEntries) == 0 {
+	queueSize := len(ap.toPurge) - 1
+	batchSize := ap.queueEntriesPerBatch
+
+	// If the queue is empty, return immediately.
+	if queueSize <= 0 {
+		ap.Unlock()
```

Comment:
Here we see manually unlocking before a return. That runs the risk we could forget to unlock.

The original code didn't worry about that because the critical section (the locked section) was tiny and obviously had no return paths.

Sometimes in more complex code we'll still do manual unlocks because it's hard to pick apart. However, whenever possible we should design our code so we can `defer foo.Unlock()`. Defer is a great facility that lets us ensure we _always_ unlock.

In this case I would factor out a small function `takeBatch` that does `ap.Lock()`, `defer ap.Unlock()` right at the top, modifies `ap.toPurge`, and returns a slice of taken objects.